spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,233 @@
+	}
+
+	type ReactorState = {
+		// Peers that have been initialized but not yet removed.
+		// The reactor should expect receiving messages from them.
+		peers: Set[NodeID],
+
+		// The reactor runs multiple routines.
+		routines: Set[Routine],
+
+		// Service implementation.
+		started: bool,
+		stopped: bool,
+
+		// Indicates whether a switch was set: reactor registration.
+		switch: bool,
+	}
+
+	var state: ReactorState
+
+	/*
+	 * GetChannels returns the list of MConnection.ChannelDescriptor. Make sure
+	 * that each ID is unique across all the reactors added to the switch.
+	 */
+	// TODO: we need a way to configure the returned channels!
+	def GetChannels(s: ReactorState): Set[ChannelDescriptor] = {
+		Set({ID: 3, Priority: 1})
+		.union(Set({ID: 7, Priority: 2}))
+		//.union(Set({ID: 3, Priority: 2})) // registration should fail!
+	}
+
+	/*
+	 * SetSwitch allows setting a switch.
+	 */
+	def SetSwitch(s: ReactorState, switch: bool): ReactorState = {
+		s.with("switch", switch)
+	}
+
+	// Code: Switch.AddReactor(name string, reactor Reactor)
+	action register() = all {
+		not(state.switch),
+		val channels = state.GetChannels()
+		val chIDs = channels.map(c => c.ID)
+		channels.size() == chIDs.size(), // channel IDs are unique
+		state' = state.SetSwitch(true),
+	}
+
+	/*
+	 * Start the service.
+	 * If it's already started or stopped, will return an error.
+	 */
+	def OnStart(s: ReactorState): ReactorState = {
+		s.with("started", true)
+	}
+
+	// Code: Switch.OnStart()
+	action start(): bool = all {
+		state.switch,
+		not(state.started) and not(state.stopped),
+		state' = state.OnStart()
+	}
+
+	/*
+	 * Stop the service.
+	 * If it's already stopped, will return an error.
+	 */
+	def OnStop(s: ReactorState): ReactorState = {
+		s.with("stopped", true)
+	}
+
+	// Code: Switch.OnStop()
+	action stop(): bool = all {
+		state.started and not(state.stopped),
+		// Either no peer was added or all peers were removed
+		state.peers.size() == 0,
+		state' = state.OnStop()
+	}
+
+	/*
+	 * Return true if the service is running.
+	 */
+	def IsRunning(s: ReactorState): bool = {
+		s.started and not(s.stopped)
+	}
+
+	/*
+	 * InitPeer is called by the switch before the peer is started. Use it to
+	 * initialize data for the peer (e.g. peer state).
+	 */
+	def InitPeer(s: ReactorState, peer: Peer): (ReactorState, Peer) = {
+		val updatedPeer = peer // Peer can be updated here
+		(s.with("peers", s.peers.union(Set(peer.ID))), updatedPeer)
+	}
+
+	// Code: Switch.addPeer(p Peer): preamble
+	action initPeer(peer: Peer): bool = all {
+		state.switch and state.IsRunning(),
+		not(peer.ID.in(state.peers)),
+		state' = state.InitPeer(peer)._1 // Updated peer not used here
+	}
+
+	/* 
+	 * AddPeer is called by the switch after the peer is added and successfully
+	 * started. Use it to start goroutines communicating with the peer.
+	 */
+	def AddPeer(s: ReactorState, peer: Peer): ReactorState = {
+		val startedRoutines = Set( {name: "ioRoutine", peer: peer} )
```

Comment:
Is this just an example, or does each reactor has this 'ioRoutine'

Reply:
This is an example. Should I point this out or maybe change the name?

Reply:
Moreover, should `AddPeer` be `pure def` as well?

Reply:
yes, add a comment that says this is an example. yes. should be a pure def :-)

Reply:
Added a comment (new commits) to clarify.